### PR TITLE
fix: revert required portal version for file chooser dialogs

### PR DESCRIPTION
**Pull request by deepak1556**
#### Description of Change

1) Revert the version bump to `3`, to help address https://github.com/electron/electron/issues/43819
2) Provide `xdg-portal-required-version` flag for application developers to control the portal usage (ex: requires support of `current_folder`)
3) If portal code path is used and a call to dialog api with `default_path` is performed then log a warning pointing to the above flag.

#### Release Notes

Notes: fix file chooser dialogs for flaptak applications

### CODE REVIEW

**Comment by ckerr** [comment]
Need to tweak the lambda capture a little but otherwise lgtm


https://productionresultssa19.blob.core.windows.net/actions-results/eed561ac-2834-415d-8c4c-e0a5cfc34f6b/workflow-job-run-413b9e83-5cfd-5774-a0bf-7a0130566c6f/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-10-28T21%3A15%3A21Z&sig=DLF5yvwJ9jJXMZehBTYW3%2FS2bsuTN1MQ56JiovqzZlI%3D&ske=2024-10-29T08%3A53%3A12Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-10-28T20%3A53%3A12Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-08-04&sp=r&spr=https&sr=b&st=2024-10-28T21%3A05%3A16Z&sv=2024-08-04

```
2024-10-28T16:06:16.0232728Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:304:26: error: variable 'xdg_portal_required_version' cannot be implicitly captured in a lambda with no capture-default specified
2024-10-28T16:06:16.0234398Z   304 |                          xdg_portal_required_version) {
2024-10-28T16:06:16.0235038Z       |                          ^
2024-10-28T16:06:16.0236274Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:279:18: note: 'xdg_portal_required_version' declared here
2024-10-28T16:06:16.0237693Z   279 |     unsigned int xdg_portal_required_version) {
2024-10-28T16:06:16.0238343Z       |                  ^
2024-10-28T16:06:16.0239107Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:11: note: lambda expression begins here
2024-10-28T16:06:16.0257715Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0258347Z       |           ^
2024-10-28T16:06:16.0259550Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by value
2024-10-28T16:06:16.0260754Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0261332Z       |            ^
2024-10-28T16:06:16.0261795Z       |            xdg_portal_required_version
2024-10-28T16:06:16.0263318Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by reference
2024-10-28T16:06:16.0264786Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0265383Z       |            ^
2024-10-28T16:06:16.0265861Z       |            &xdg_portal_required_version
2024-10-28T16:06:16.0266833Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by value
2024-10-28T16:06:16.0267775Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0268299Z       |            ^
2024-10-28T16:06:16.0268662Z       |            =
2024-10-28T16:06:16.0269428Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by reference
2024-10-28T16:06:16.0270436Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0271011Z       |            ^
2024-10-28T16:06:16.0271410Z       |            &
2024-10-28T16:06:16.0273104Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:311:68: error: variable 'xdg_portal_required_version' cannot be implicitly captured in a lambda with no capture-default specified
2024-10-28T16:06:16.0275101Z   311 |             VLOG(1) << "File chooser portal expected version: " << xdg_portal_required_version;
2024-10-28T16:06:16.0276082Z       |                                                                    ^
2024-10-28T16:06:16.0277430Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:279:18: note: 'xdg_portal_required_version' declared here
2024-10-28T16:06:16.0278864Z   279 |     unsigned int xdg_portal_required_version) {
2024-10-28T16:06:16.0279470Z       |                  ^
2024-10-28T16:06:16.0280276Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:11: note: lambda expression begins here
2024-10-28T16:06:16.0281219Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0281756Z       |           ^
2024-10-28T16:06:16.0282850Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by value
2024-10-28T16:06:16.0283949Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0284514Z       |            ^
2024-10-28T16:06:16.0284949Z       |            xdg_portal_required_version
2024-10-28T16:06:16.0286220Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by reference
2024-10-28T16:06:16.0287410Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0287985Z       |            ^
2024-10-28T16:06:16.0288440Z       |            &xdg_portal_required_version
2024-10-28T16:06:16.0289419Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by value
2024-10-28T16:06:16.0290697Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0291290Z       |            ^
2024-10-28T16:06:16.0291669Z       |            =
2024-10-28T16:06:16.0292501Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by reference
2024-10-28T16:06:16.0293563Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0294133Z       |            ^
2024-10-28T16:06:16.0294535Z       |            &
2024-10-28T16:06:16.0294919Z 2 errors generated.
```

**Review by ckerr — DISMISSED** [review]
Need to tweak the lambda capture to fix a the build, but otherwise looks good



```
2024-10-28T16:06:16.0232728Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:304:26: error: variable 'xdg_portal_required_version' cannot be implicitly captured in a lambda with no capture-default specified
2024-10-28T16:06:16.0234398Z   304 |                          xdg_portal_required_version) {
2024-10-28T16:06:16.0235038Z       |                          ^
2024-10-28T16:06:16.0236274Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:279:18: note: 'xdg_portal_required_version' declared here
2024-10-28T16:06:16.0237693Z   279 |     unsigned int xdg_portal_required_version) {
2024-10-28T16:06:16.0238343Z       |                  ^
2024-10-28T16:06:16.0239107Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:11: note: lambda expression begins here
2024-10-28T16:06:16.0257715Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0258347Z       |           ^
2024-10-28T16:06:16.0259550Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by value
2024-10-28T16:06:16.0260754Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0261332Z       |            ^
2024-10-28T16:06:16.0261795Z       |            xdg_portal_required_version
2024-10-28T16:06:16.0263318Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by reference
2024-10-28T16:06:16.0264786Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0265383Z       |            ^
2024-10-28T16:06:16.0265861Z       |            &xdg_portal_required_version
2024-10-28T16:06:16.0266833Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by value
2024-10-28T16:06:16.0267775Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0268299Z       |            ^
2024-10-28T16:06:16.0268662Z       |            =
2024-10-28T16:06:16.0269428Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by reference
2024-10-28T16:06:16.0270436Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0271011Z       |            ^
2024-10-28T16:06:16.0271410Z       |            &
2024-10-28T16:06:16.0273104Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:311:68: error: variable 'xdg_portal_required_version' cannot be implicitly captured in a lambda with no capture-default specified
2024-10-28T16:06:16.0275101Z   311 |             VLOG(1) << "File chooser portal expected version: " << xdg_portal_required_version;
2024-10-28T16:06:16.0276082Z       |                                                                    ^
2024-10-28T16:06:16.0277430Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:279:18: note: 'xdg_portal_required_version' declared here
2024-10-28T16:06:16.0278864Z   279 |     unsigned int xdg_portal_required_version) {
2024-10-28T16:06:16.0279470Z       |                  ^
2024-10-28T16:06:16.0280276Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:11: note: lambda expression begins here
2024-10-28T16:06:16.0281219Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0281756Z       |           ^
2024-10-28T16:06:16.0282850Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by value
2024-10-28T16:06:16.0283949Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0284514Z       |            ^
2024-10-28T16:06:16.0284949Z       |            xdg_portal_required_version
2024-10-28T16:06:16.0286220Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: capture 'xdg_portal_required_version' by reference
2024-10-28T16:06:16.0287410Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0287985Z       |            ^
2024-10-28T16:06:16.0288440Z       |            &xdg_portal_required_version
2024-10-28T16:06:16.0289419Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by value
2024-10-28T16:06:16.0290697Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0291290Z       |            ^
2024-10-28T16:06:16.0291669Z       |            =
2024-10-28T16:06:16.0292501Z ../../ui/shell_dialogs/select_file_dialog_linux_portal.cc:291:12: note: default capture by reference
2024-10-28T16:06:16.0293563Z   291 |           [](scoped_refptr<dbus::Bus> bus,
2024-10-28T16:06:16.0294133Z       |            ^
2024-10-28T16:06:16.0294535Z       |            &
2024-10-28T16:06:16.0294919Z 2 errors generated.
```

File: docs/api/command-line-switches.md
Comment (samuelmaddock) [inline docs/api/command-line-switches.md:248]: Who is the intended consumer for this switch? Application developers, users, or both? For the recent bump to v4, is this so application users can switch back the behavior if they require it?

I also wonder if we should consider reverting the default to v3. #43570 contained an undocument breaking change in behavior that applications may not be expecting.

Comment (VerteDinde) [reply docs/api/command-line-switches.md:248]: I don't think we should revert the default, since it looks like the change was accepted upstream (see https://chromium-review.googlesource.com/c/chromium/src/+/5786982), but we probably should document this as a breaking change, at least retroactively, so people know

**Comment by deepak1556** [comment]
Sorry for the delay, I was testing some idea to see if we can repurpose the function `SelectFileDialogLinuxPortal::CheckPortalAvailabilityOnBusThread` to not only check for portal availability but also ensure the option `current_folder` is available on the relevant interfaces but I couldn't find a way to probe it without making the actual method call.

Rethinking this PR, based on the conversations from https://github.com/electron/electron/issues/43819 it seems the affected  scenarios with the portal version bump is harder to recover (generic filechooser inside the flatpak application is opened without read/write access to any file on the host) than the scenario with `defaultPath` https://github.com/electron/electron/issues/43310. Thoughts on the following next steps,

1) Revert the version bump to `3`, to help address https://github.com/electron/electron/issues/43819
2) Provide `xdg-portal-required-version` flag for application developers to control the portal usage (ex: requires support of `current_folder`)
3) If portal code path is used and a call to dialog api with `default_path` is performed then log a warning pointing to the above flag.

File: docs/api/command-line-switches.md
Comment (codebytere) [reply docs/api/command-line-switches.md:248]: @VerteDinde I do think (in agreement w/ deepak below)  that we should change back the default, since as of now v4 is so rarely available that every app might be required to float this flag for potentially years.

**Review by jkleinsc — COMMENTED** [review]
API LGTM

File: docs/api/command-line-switches.md
Comment (samuelmaddock) [inline docs/api/command-line-switches.md:248]: What's the behavior when the environment doesn't meet the required version?

**Review by samuelmaddock — APPROVED** [review]
API LGTM

Have a question for clarity, but won't block merging this.

File: patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch
Comment (itsananderson) [inline patches/chromium/feat_add_support_for_missing_dialog_features_to_shell_dialogs.patch:212]: There's a [suggestion in the GH issue](https://github.com/electron/electron/issues/43819#issuecomment-2448475826) about querying for the available portal version rather than hardcoding to 3 or 4. Would that be a viable approach?

File: docs/api/command-line-switches.md
Comment (deepak1556) [reply docs/api/command-line-switches.md:248]: Missed documenting here, we fallback to the gtk or kde based file dialogs depending on the environment.

File: docs/api/command-line-switches.md
Comment (jkleinsc) [inline docs/api/command-line-switches.md:247]:
```suggestion
in order to use the portal backend for file dialogs on linux. File dialogs
```

File: docs/api/dialog.md
Comment (jkleinsc) [inline docs/api/dialog.md:84]: Could we be more explicit here, something like:
```suggestion
**Note:** On Linux `defaultPath` is not supported when using portal file chooser
dialogs unless the portal backend is version 4 or higher. You can use `--xdg-portal-required-version`
[command-line switch](./command-line-switches.md#--xdg-portal-required-versionversion)
to force gtk or kde dialogs.
```

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Overall, LGTM, just a small grammar correction and I also wonder if we should be more explicit about explaining that the portal version needs to be 4 or greater to use `defaultPath`.

File: docs/api/dialog.md
Comment (codebytere) [inline docs/api/dialog.md:194]: `defaultPath` is only unsupported on versions <4 for open dialogs - save dialogs work as expected.

File: docs/api/dialog.md
Comment (codebytere) [inline docs/api/dialog.md:236]: `defaultPath` is only unsupported on versions <4 for open dialogs - save dialogs work as expected.

**Review by codebytere — CHANGES_REQUESTED** [review]

File: docs/api/dialog.md
Comment (deepak1556) [reply docs/api/dialog.md]: Thanks for catching.

For reference, https://github.com/flatpak/xdg-desktop-portal/issues/796

**Review by codebytere — APPROVED** [review]

File: docs/api/command-line-switches.md
Comment (jkleinsc) [inline docs/api/command-line-switches.md:248]:
```suggestion
will fallback to using gtk or kde depending on the desktop environment when
```

**Review by jkleinsc — CHANGES_REQUESTED** [review]
One more grammar nit.

**Review by jkleinsc — APPROVED** [review]

**Comment by razzeee** [comment]
Testing with electron fiddle, this seems to be missing from 35? Is that possible?

**Comment by deepak1556** [comment]
The changes are present but there could be a regression to the patch from a chromium version bump since portal code path was refactored in upstream. Can you open an issue with a repro and we can take a look into it.

**Comment by razzeee** [comment]
@deepak1556 I've opened that repro issue here https://github.com/electron/electron/issues/46652

**Pull request merged**
